src/commons/utils/RequestHelper.tsx

```diff
@@ -0,0 +1,173 @@
+
+/**
+ * @returns {(Response|null)} Response if successful, otherwise null.
+ *
+ * If response status is 401 Unauthorized, this function will attempt to
+ * refresh the JWT tokens before trying the API call again.
+ *
+ * If user is still unauthorized after refresh token flow:
+ * - User in non-AssessmentWorkspace page: perform force logout
+ * - User in AssessmentWorkspace page: prompt the user to manually save their work; does not force logout
+ *
+ * NOTE: If multiple API calls are made with an expired access token, only one refresh token flow will be made.
+ * All these API calls will then be retried thereafter.
+ */
+export const request = async (
+  path: string,
+  method: RequestMethod,
+  opts: RequestOptions
+): Promise<Response | null> => {
+  const fetchOptions = generateApiCallHeadersAndFetchOptions(method, opts);
+
+  try {
+    const resp = await fetch(`${Constants.backendUrl}/v2/${path}`, fetchOptions);
+    if (resp.ok) {
+      return resp;
+    }
+
+    // Non-unauthorized errors (these don't need to trigger refresh token flow)
+    if (resp.status !== 401) {
+      showWarningMessage(opts.errorMessage ? opts.errorMessage : getResponseErrorMessage(resp));
+      return null;
+    }
+
+    // Unauthorized errors (trigger refresh token flow)
+    if (!opts.refreshToken) {
+      // If no refreshToken passed in, we are in `postRefresh` itself AND the refresh token flow failed
+      // (i.e. invalid refresh token sent)
+      return null;
+    }
+
+    // Perform refresh token flow before retrying API call.
+    // If refresh token flow or retried API call 401s again, determine whether to force logout.
+    try {
+      if (!refreshingTokensPromise) {
+        refreshingTokensPromise = postRefresh(opts.refreshToken);
+      }
+
+      // Blocks other API calls that failed due to expired access token, while performing a single refresh token flow
+      const newTokens = await refreshingTokensPromise;
+      if (!newTokens) {
+        throw Error('Invalid refreshed token');
+      }
+
+      store.dispatch(actions.setTokens(newTokens));
+      const updatedFetchOptions = _.cloneDeep(fetchOptions);
+      updatedFetchOptions.headers.set('Authorization', `Bearer ${newTokens.accessToken}`);
+
+      const retriedResp = await fetch(`${Constants.backendUrl}/v2/${path}`, updatedFetchOptions);
+      if (retriedResp.ok) {
+        return retriedResp;
+      }
+
+      if (retriedResp.status === 401) {
+        // Something went wrong with the refreshed access token (this should never happen)
+        throw Error('Invalid refreshed access token');
+      }
+
+      showWarningMessage(
+        opts.errorMessage ? opts.errorMessage : getResponseErrorMessage(retriedResp)
+      );
+      return null;
+    } catch (err) {
+      // Refresh token flow or retried API call 401s again. Force logout in non-AssessmentWorkspace routes.
+      const assessmentPathRegex = /\/courses\/\d+\/[a-zA-Z]+\/\d+\/\d+/;
```

Comment:
This might unintentionally result in bugs if in the future, the paths for the workspaces are changed for some reason, as those who changed it might be unaware for the existence of this regex. Would it be better if instead, this is exported from, perhaps, where we define the paths?